### DISCUSSION TITLE
[@types/jest] Type deduction fails in `it.each` when parts of the test data are ignored using typescript comments

### DISCUSSION THREAD

**Discussion by Fullfungo** [DefinitelyTyped/DefinitelyTyped#71498] [Issues with a @types package]
### Version

29.5.14

### Steps to reproduce

1. Clone the repo https://github.com/SmolPatat/jest-type-bug/tree/main
2. Follow the README.
   1. Run `npm install`
   2. Run `npm run validate`
3. TypeScript will throw an error.

### Example code
```ts
type VeryComplex = {
    field1: number,
    manyOtherFields: unknown
};

it.each<{ input: VeryComplex, expected: number }>([
    {
        // @ts-expect-error -- Missing irrelevant fields.
        input: {
            field1: 42
        },
        expected: 42
    }
])('type issue', ({ input, expected }) => {
    expect(input).toHaveProperty('field1', expected)
})
```

### Expected behavior

Since I explicitly specified the type in `it.each`, this exact type should be used for the arguments of the test function, no matter what I do with the test data.

### Actual behavior

Jest deduces some goofy types, even though it doesn't need to deduce anything, since it has the type explicitly provided to it. The types deduced are not valid and suggest dangerous changes that will make it fail at runtime.

### Additional context

Omitting some object fields in the test data is a common practice. It should be possible to use the recommended solutions such as `@ts-expect-error` or `@ts-ignore` when the developer wants to only provide the data relevant to the test.

[a comment by an automated account is not shown]

**Comment by mrazauskas** [comment]
I think the behaviour is correct. Types here and many other places of `@types/jest` are implemented so that:

- when a type argument is provided, it is used to make sure that the value matches shape of that type.

For instance: 

```ts
expect(actual).toBe<{ a: string }>(expected);
```

Here `expected` must be of expected type. That is similar to what you see with `it.each()`. If you have to widen the inferred type, use type casting.

**Reply by Fullfungo** [reply]
But then `bug-2.test.ts` is still problematic. It does not raise any type errors, which makes even less sense than the original bug. Additionally, it deduces the types to be `input: VeryComplex` and `expected: number` instead of `any` that you would expect in case of a type error. This can't be right.

```ts
type VeryComplex = {
    field1: number,
    manyOtherFields: unknown
};

it.each<{ input: VeryComplex, expected: number }>([
    {
        // @ts-expect-error -- Missing irrelevant fields.
        input: {
            field1: 42
        },
        expected: 42
    }
])('no type issue', (...{ input, expected }) => {
    expect(input).toHaveProperty('field1', expected)
})
```

**Comment by mrazauskas** [comment]
As I wrote: if you have to widen the inferred type, use type casting.

```ts
type VeryComplex3 = {
  field1: number;
  manyOtherFields: unknown;
};

it.each([
  {
    input: {
      field1: 42,
    },
    expected: 42,
  } as { input: VeryComplex3; expected: number },
])("no type issue", ({ input, expected }) => {
  expect(input).toHaveProperty("field1", expected);
});
```

Also don’t assume `// @ts-expect-error` will produce the `any` type. If you need `any`, just use `as any`.

**Reply by Fullfungo** [reply]
Thanks for the suggestion.

Btw, I didn't mean that `// @ts-expect-error` will necessarily produce `any`. I meant something like the following code
```ts
function F<T>(data: { idk: T }): (arg: T) => T {
  return x=>x
}

const x = {idk: 3};
const y = F<boolean>(
  // @ts-expect-error
  x
);
```
would result in `y: (arg: boolean) => boolean` (as it does in normal TS) instead of a broken type. Maybe I'm missing something, but I would expect Jest to work this way.

**Reply by Fullfungo** [reply]
Also, jest is resolving it as `T extends any[] | [any]` for some reason, which is why it suggests `...` on the arguments.
